crates/oxc_codegen/src/lib.rs

```diff
@@ -377,6 +428,154 @@ impl<const MINIFY: bool> Codegen<MINIFY> {
+        self.internal_add_source_mapping(position, Some(name));
+    }
+
+    fn internal_add_source_mapping(&mut self, position: u32, name: Option<&str>) {
+        if self.enable_sourcemap {
+            if matches!(self.last_position, Some(last_position) if last_position == position) {
+                return;
+            }
+            let (original_line, original_column) = self.search_original_line_and_column(position);
+            self.update_generated_line_and_column();
+            // println!("span {}, {:?} => {:?}", position, (original_line, original_column), (self.generated_line, self.generated_column));
+            let name_id = name.map(|s| self.sourcemap_builder.add_name(s));
+            self.sourcemap_builder.add_raw(
+                self.generated_line,
+                self.generated_column,
+                original_line,
+                original_column,
+                Some(self.source_id),
+                name_id,
+            );
+            self.last_position = Some(position);
+        }
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
+    fn search_original_line_and_column(&self, position: u32) -> (u32, u32) {
+        let result = self
+            .line_offset_tables
+            .partition_point(|table| table.byte_offset_to_start_of_line <= position as usize);
+        let original_line = if result > 0 { result - 1 } else { 0 };
+        let line = &self.line_offset_tables[original_line];
+        let mut original_column = (position as usize) - line.byte_offset_to_start_of_line;
+        if original_column >= line.byte_offset_to_first {
+            if let Some(cols) = &line.columns {
+                original_column = cols[original_column - line.byte_offset_to_first];
+            }
+        }
+        (original_line as u32, original_column as u32)
+    }
+
+    fn update_generated_line_and_column(&mut self) {
+        // SAFETY: criteria of `from_utf8_unchecked`.are met.
+        let s = unsafe { std::str::from_utf8_unchecked(&self.code[self.last_generated_update..]) };
+        // println!("\n---- {:?}", s);
+        for (i, ch) in s.chars().enumerate() {
+            match ch {
+                '\r' | '\n' | LS | PS => {
+                    // Handle Windows-specific "\r\n" newlines
+                    if ch == '\r' && self.code[self.last_generated_update + i + 1] == b'\n' {
+                        continue;
+                    }
+                    self.generated_line += 1;
+                    self.generated_column = 0;
+                }
+                _ => {
+                    if ch as u32 <= 0xFF {
+                        self.generated_column += 1;
+                    } else {
+                        self.generated_column += 2;
+                    }
+                }
+            }
+        }
+        self.last_generated_update = self.code.len();
+    }
+
+    fn generate_line_offset_tables(content: &str) -> Vec<LineOffsetTable> {
```

Comment:
I think we can optimize this with `memchr`, but we'll need some benchmark setup before this.